### PR TITLE
Always resume automatic sessions on remote (BugFix)

### PR DESCRIPTION
**Pull request by kissiel**
## Description

This PR changes how Checkbox handles automatic session crashes when using Checkbox Agent.

Previously if a job crashed or hang, there was no way to continue the session - when the Controller reconnected to the respawned Agent, it started the session afresh.

With the change I'm proposing here, the Checkbox Agent always resumes the automatic (silent) sessions. The previous job's outcome is decided depending on the existence of the `noreturn` flag. But the session **always** moves forward. It means that the sesion should never just disappear or keep rerunning the same job.

## Resolved issues

Fixes: https://github.com/canonical/checkbox/issues/22 
Fixes: https://github.com/canonical/checkbox/issues/722
Fixes: [CHECKBOX-796](https://warthogs.atlassian.net/jira/software/c/projects/CHECKBOX/boards/590?selectedIssue=CHECKBOX-796)
Fixes: [CHECKBOX-867](https://warthogs.atlassian.net/jira/software/c/projects/CHECKBOX/boards/590?selectedIssue=CHECKBOX-867)

## Documentation
I'm providing description of the agent bootup together with a flow chart.
## Tests
I've covered the code with unit tests and I'm proposing a Metabox scenario here.

[CHECKBOX-796]: https://warthogs.atlassian.net/browse/CHECKBOX-796?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CHECKBOX-867]: https://warthogs.atlassian.net/browse/CHECKBOX-867?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by Hook25 — COMMENTED** [review]
A few preliminary comments here and there

File: checkbox-ng/checkbox_ng/launcher/agent.py [hunk 1]
```diff
@@ -91,76 +97,95 @@ class RemoteAgent():
...
         self._server.start()
 
     def register_arguments(self, parser):
-        parser.add_argument('--resume', action='store_true', help=_(
-            "resume last session"))
-        parser.add_argument('--port', type=int, default=18871, help=_(
-            "port to listen on"))
+        parser.add_argument(
+            "--resume", action="store_true", help=_("resume last session")
+        )
```
[the remark this answers is not in the document]
Comment (kissiel) [reply checkbox-ng/checkbox_ng/launcher/agent.py:171]: Nope. This becomes dead and there for historical reasons. Agent from now on **ALWAYS** resumes.

File: checkbox-ng/checkbox_ng/launcher/agent.py [hunk 2]
```diff
@@ -91,76 +97,95 @@ class RemoteAgent():
...
+
+    To determine that we need to take the original launcher that had been used
+    when the session was started, recreate it as a proper Launcher object, and
+    check if it's in fact non-interactive.
+    """
+    # app blob is a bytes string with a utf-8 encoded json
+    # let's decode it and parse it as json
+    app_blob = json.loads(resumable_session.metadata.app_blob.decode("utf-8"))
+    launcher = Configuration.from_text(app_blob["launcher"], "resumed session")
+    return launcher.sections["ui"].get("type") == "silent"
```
[the remark this answers is not in the document]
Comment (kissiel) [reply checkbox-ng/checkbox_ng/launcher/agent.py:176]: For the previous session we store it in that app_blob. So this is the place.
As for the peeker being used, it is. The info about resumable sessions that we pull when the agent is starting are indeed just the peeks. 

As for the structure.
We could have a parametric "get resumable sessions", where there could be a way to filter out the sessions we're looking for, but that would result in more API surface. Now we get all the peeks and decide in the application code (local or remote).

File: checkbox-ng/plainbox/impl/session/agent_bootup.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+graph TD;
+  proc("agent process starts")
+  load("load previous session")
+  resume("resume the previous session")
+  resume_crashed("resume the previous session")
+  auto_session{"was the previous session non-interactive"}
+  mark_pass("mark last running job as passing")
+  mark_crash("mark last running job as crashing")
+  idle("go into idle state")
+  listen("listen for the controllers")
```
[the remark this answers is not in the document]
Comment (kissiel) [reply checkbox-ng/plainbox/impl/session/agent_bootup.md]: Fixed!

File: checkbox-ng/plainbox/impl/session/agent_bootup.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  proc --> load
+  last_job{"was the last job a `noreturn` job?"}
+  load -->last_job
+  last_job-->|yes| resume
+  resume --> mark_pass
+  
+  last_job-->|no| auto_session
+  auto_session-->|no| idle
```
[the remark this answers is not in the document]
Comment (kissiel) [reply checkbox-ng/plainbox/impl/session/agent_bootup.md]: That's a good question :)

File: checkbox-ng/plainbox/impl/session/agent_bootup.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  idle --> listen
+  mark_pass --> listen
+
+  auto_session-->|yes| resume_crashed
+  resume_crashed --> mark_crash
+  mark_crash --> listen
+
+  
+
+
```
[the remark this answers is not in the document]
Comment (kissiel) [reply checkbox-ng/plainbox/impl/session/agent_bootup.md]: The file was temporary. It is removed completely now.

File: checkbox-ng/plainbox/impl/session/agent_bootup.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# Checkbox Agent bootup process
+
+
+```mermaid
+graph TD;
+  proc("agent process starts")
+  load("load previous session")
+  resume("resume the previous session")
+  resume_crashed("resume the previous session")
+  auto_session{"was the previous session non-interactive"}
```
[the remark this answers is not in the document]
Comment (kissiel) [reply checkbox-ng/plainbox/impl/session/agent_bootup.md]: Replaced the double negatives with simpler terms and descriptions.

File: checkbox-ng/plainbox/impl/session/agent_bootup.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  mark_pass("mark last running job as passing")
+  mark_crash("mark last running job as crashing")
+  idle("go into idle state")
+  listen("listen for the controllers")
+  proc --> load
+  last_job{"was the last job a `noreturn` job?"}
```
[the remark this answers is not in the document]
Comment (kissiel) [reply checkbox-ng/plainbox/impl/session/agent_bootup.md]: Good tip, thanks!

File: checkbox-ng/checkbox_ng/launcher/agent.py [hunk 8]
```diff
@@ -91,76 +97,95 @@ class RemoteAgent():
...
         if not is_passwordless_sudo():
             raise SystemExit(
-                _("System is not configured to run sudo without a password!"))
+                _("System is not configured to run sudo without a password!")
+            )
+        if ctx.args.resume:
+            msg = (
+                "--resume is deprecated and will be removed soon. "
+                "Automated sessions are now always resumed. "
+                "Manual sessions can be resumed from the welcome screen."
```
[the remark this answers is not in the document]
Comment (kissiel) [reply checkbox-ng/checkbox_ng/launcher/agent.py:113]: That even hasn't landed yet.

[the remark this answers is not in the document]
Comment (Hook25) [reply checkbox-ng/checkbox_ng/launcher/agent.py:113]: true :(

File: checkbox-ng/checkbox_ng/launcher/test_agent.py [hunk 9]
```diff
@@ -61,3 +66,132 @@ def test_invoked_ok(
...
+            type = silent
+        """
+        app_blob = json.dumps({"launcher": a_non_interactive_launcher})
+        metadata = SessionMetaData(app_blob=app_blob.encode("utf-8"))
+
+        candidate = ResumeCandidate("an_id", metadata)
+        self.assertTrue(is_the_session_noninteractive(candidate))
+
+    def test_an_interactive(self):
+        a_non_interactive_launcher = ""  # the defautl one is interactive
```
Comment (Hook25) [inline checkbox-ng/checkbox_ng/launcher/test_agent.py:160]: This also tests the default in a way, I don't think this is a good idea, just make it interactive

File: checkbox-ng/plainbox/impl/session/remote_assistant.py [hunk 10]
```diff
@@ -719,7 +719,7 @@ def resume_by_id(self, session_id=None):
 
         result_dict = {
             "outcome": IJobResult.OUTCOME_PASS,
-            "comments": _("Automatically passed after resuming execution"),
+            "comments": "Automatically passed after resuming execution",
```
Comment (Hook25) [inline checkbox-ng/plainbox/impl/session/remote_assistant.py:722]: I wish we could do this, but I feel like we should do all in one go

File: checkbox-ng/plainbox/impl/session/test_remote_assistant.py [hunk 11]
```diff
@@ -46,6 +58,215 @@ def not_allowed(self, *args):
         with self.assertRaises(AssertionError):
             not_allowed(self_mock)
 
+    @mock.patch("plainbox.impl.secure.sudo_broker.is_passwordless_sudo")
+    def test__reset_sa(self, is_passwordless_sudo_mock):
+        with mock.patch.object(SessionAssistant, "__init__") as init_mock:
```
Comment (Hook25) [inline checkbox-ng/plainbox/impl/session/test_remote_assistant.py:63]: Why do we do it this way in this function? Can't we mock above the function the SessionAssistant and check it was called?

File: docs/explanation/remote.rst [hunk 12]
```diff
@@ -92,6 +92,51 @@ End this test session preserving its data and launch a new one
...
+
+.. mermaid::
+
+  graph TD;
+
+  process("agent process starts")
+    load("load previous session")
+    resume("resume the previous session")
+    resume_crashed("resume the previous session")
+    interactive{"was the previous interactive"}
```
Comment (Hook25) [inline docs/explanation/remote.rst:118]: las session interactive?
so it is the same as last job noreturn

File: metabox/metabox/lxd_profiles/checkbox.profile
Comment (Hook25) [inline metabox/metabox/lxd_profiles/checkbox.profile:1]: Was this committed here by mistake?  this may be necessary but it has nothing to do with this PR

**Review by Hook25 — CHANGES_REQUESTED** [review]
A few minor comments, beside them this is pretty much ready to land. Thanks for adding all those tests

File: docs/explanation/remote.rst (hunk 12 above)
Comment (kissiel) [reply docs/explanation/remote.rst:118]: noreturn-ness and interactivity are orthogonal

(and I fixed the missing "session" word)

**Review by pieqq — CHANGES_REQUESTED** [review]
This is a pretty exciting PR, especially if it fixes this long-standing issue of session resuming "instability"!

Here is how I tested your branch. This is a long explanation for my future self (or anyone who wish to test this branch, really), but you can jump to the "Tests" section below.

TL;DR: I found a problem when reconnecting to the agent **after** the ongoing job is finished.

# Setup

## Checkbox controller

On my laptop, I already have a virtual environment setup for Checkbox. I just point to your branch:

```
(venv) $ git switch solve-resume-on-remote
```

I use this venv for the Checkbox controller.

## Checkbox agent

For the Checkbox agent, I create an LXC container running 22.04:

```
$ lxc launch images:ubuntu/22.04 jammy
$ lxc shell jammy
```

The rest of the commands are run in the container:

```
# apt install python3.10-venv python3-virtualenv git
# git clone https://github.com/canonical/checkbox.git
# cd checkbox/
# git switch solve-resume-on-remote
```

I follow the [Contrib guide](https://github.com/canonical/checkbox/blob/main/CONTRIBUTING.md#testing) to get Checkbox installed in a venv. In the end, checkbox-cli lives in `/root/checkbox/checkbox-ng/venv/bin/checkbox-cli` and the providers are in described in `/root/checkbox/checkbox-ng/venv/share/plainbox-providers-1`.

I put the following in `/etc/systemd/system/checkbox-ng.service`:

```
[Unit]
Description=Checkbox Remote Service
Wants=network.target

[Service]
ExecStart=/root/checkbox/checkbox-ng/venv/bin/checkbox-cli run-agent
SyslogIdentifier=checkbox-ng.service
Environment="XDG_CACHE_HOME=/var/cache/"
Environment="PROVIDERPATH=/root/checkbox/checkbox-ng/venv/share/plainbox-providers-1"
Restart=always
RestartSec=1
TimeoutStopSec=30
Type=simple

[Install]
WantedBy=multi-user.target
```

and I install the checkbox-ng service and start it:

```
# systemctl daemon-reload
# systemctl enable checkbox-ng.service
```

Now, everything is in place. I can start a remote session from the controller by running:

```
(venv) $ checkbox-cli control <IP of my lxc container>
```

## Sample jobs and test plan

In the 22.04 container, I create a new `pieq.pxu` file in `/root/checkbox/providers/base/units/` and put the following in it:

```
unit: job
id: pieq/test
command:
 for i in $(seq 1 30);
 do
     echo "Iteration $i/30..."
     sleep 1
 done
flags: simple noreturn

unit: job
id: pieq/wrapup
command:
 echo "Wrapping up..."
flags: simple

unit: test plan
id: pieq
_name: pieq
include:
    pieq/test
    pieq/wrapup
```

the `pieq/test` job will run for 30 seconds and will show the current status of the job, so it's handy to see what's going on. It has the `noreturn` flag, but of course you can remove this flag if you want to test other use cases.

I need to restart the systemd service, otherwise this test plan will not be visible to Checkbox:

```
# systemctl restart checkbox-ng.service
```

## Launcher

In order to simulate a non-interactive test run, I create the following launcher file (`pieq.launcher`):

```
[launcher]
launcher_version = 1
app_id = com.canonical.certification:PR859
stock_reports = text

[test plan]
unit = com.canonical.certification::pieq
forced = yes

[test selection]
forced = yes

[ui]
type = silent

[transport:outfile]
type = stream
stream = stdout

[exporter:text]
unit = com.canonical.plainbox::text

[report:screen]
transport = outfile
exporter = text
```

To run it from the controller side with:

```
(venv) $ checkbox-cli control <IP of my lxc container> pieq.launcher
```

# Tests


## Resuming a non-interactive session after simulating a DUT crash :heavy_check_mark: 

1. Run Checkbox remote using the launcher, which starts `pieq/test` (which runs for 30 seconds): 

```
(venv) $ checkbox-cli control <IP of my lxc container> pieq.launcher
```

→ The test starts running

2. Simulate a DUT crash by stopping and starting the LXC container:

```
lxc stop jammy && lxc start jammy
```

→ the session resumes, the test is marked as passed and the test run proceeds to the next test before finalizing the session:

```
Connecting to 10.146.223.75:18871. Timeout: 600s
-----------------------------[ Running job 1 / 2 ]------------------------------
---------------------------------[ pieq/test ]----------------------------------
ID: com.canonical.certification::pieq/test
Category: Uncategorised
--------------------------------------------------------------------------------
Iteration 1/30...
Iteration 2/30...
Iteration 3/30...
Iteration 4/30...
Iteration 5/30...
Iteration 6/30...
Iteration 7/30...
Iteration 8/30...
Connection lost!
connection closed by peer
Reconnecting ...
Reconnected (took: 0s)
---------------------------------[ pieq/test ]----------------------------------
ID: com.canonical.certification::pieq/test
Category: Uncategorised
--------------------------------------------------------------------------------
Outcome: job passed
-----------------------------[ Running job 2 / 2 ]------------------------------
--------------------------------[ pieq/wrapup ]---------------------------------
ID: com.canonical.certification::pieq/wrapup
Category: Uncategorised
--------------------------------------------------------------------------------
Wrapping up...
--------------------------------------------------------------------------------
Outcome: job passed
==================================[ Results ]===================================
  job passed   : pieq/test
  job passed   : pieq/wrapup
```

If the job `pieq/test` does not have the `noreturn` flag, the only difference is that the job will be marked as "crashed" when resuming the session.

→ All good!

## Reconnecting to agent after the controller stopped/crashed :x: 

One of the issue this should fix is #22 , which mentions

> While testing is ongoing, restart your host computer.

So:

1. Run Checkbox remote using the launcher, which starts `pieq/test` (which runs for 30 seconds): 

```
(venv) $ checkbox-cli control <IP of my lxc container> pieq.launcher
```

→ The test starts running

2. Close the terminal where the controller is running. Wait for 30 seconds, then try reconnecting to the agent:

```
(venv) $ checkbox-cli control 10.146.223.75
$PROVIDERPATH is defined, so following provider sources are ignored ['/usr/local/share/plainbox-providers-1', '/usr/share/plainbox-providers-1', '/home/pieq/.local/share/plainbox-providers-1', '/var/tmp/checkbox-providers-develop'] 
Connecting to 10.146.223.75:18871. Timeout: 600s
Rejoined session.
In progress: com.canonical.certification::pieq/test (1/2)
Iteration 17/30...
Iteration 18/30...
Iteration 19/30...
Iteration 20/30...
Iteration 21/30...
Iteration 22/30...
Iteration 23/30...
Iteration 24/30...
Iteration 25/30...
Iteration 26/30...
Iteration 27/30...
Iteration 28/30...
Iteration 29/30...
Iteration 30/30...
```

aaaaaaaaand nothing happens. The session never goes on to the next job (`pieq/wrapup`), and never finishes. This is because the job has finished running by the time we reconnect to the agent.

## Running the previous tests using an interactive session :x: 

If I try to do the same as before, but with an interactive session (that is, without providing the launcher and instead running `checkbox-cli control <IP of my lxc container>` and selecting the test plan and the tests to run, as commonly done by QA team), the session is never resumed as expected, and instead I am back to the "Select test plan" screen.

This is probably what [the documentation explains](https://canonical-checkbox--859.com.readthedocs.build/en/859/explanation/remote.html#automatic-session-resume):

> Otherwise, it waits for a Controller to connect and chose what to do.

except the "what to do" just means "forget the on-going session and start a new one". As I understand, there is still some work by @Hook25 to get this working.

File: docs/explanation/remote.rst (hunk 12 above)
Comment (Hook25) [reply docs/explanation/remote.rst:118]: I ment so the two lines are similar, not related. My objective is to make the diamonds also a little bit less HUGE because every char you add to them they seem to grow by 2m^2

**Comment by kissiel** [comment]
@pieqq
 from the 3 experiments you've done, the middle one is about controller bugs more than what this PR is about.

For the third one, there's another set of stories that ought to land this week that solve that.

This PR is about reliable automated runs.

File: checkbox-ng/checkbox_ng/launcher/agent.py [hunk 13]
```diff
@@ -91,76 +97,105 @@ class RemoteAgent():
...
+    Check if given session is non-interactive.
+
+    To determine that we need to take the original launcher that had been used
+    when the session was started, recreate it as a proper Launcher object, and
+    check if it's in fact non-interactive.
+    """
+    # app blob is a bytes string with a utf-8 encoded json
+    # let's decode it and parse it as json
+    app_blob = json.loads(resumable_session.metadata.app_blob.decode("utf-8"))
+    launcher = Configuration.from_text(app_blob["launcher"], "resumed session")
```
Comment (pieqq) [inline checkbox-ng/checkbox_ng/launcher/agent.py:179]: You may want to check for the existence of the `launcher` key here. If a session is started manually without using a launcher, the session will not contain a launcher.

I just tested this by:

1. running checkbox-cli and starting a test run, then save & quitting
2. then:

```python
>>> sa = SessionAssistant()
>>> sessions = list(sa.get_resumable_sessions())
>>> app_blob = json.loads(sessions[0].metadata.app_blob.decode("utf-8"))
>>> app_blob
{'testplan_id': 'com.canonical.certification::client-cert-desktop-22-04-manual',
 'description': ''}

>>> Configuration.from_text(app_blob["launcher"], "resumed session")
KeyError: 'launcher'

```

File: checkbox-ng/checkbox_ng/launcher/checkbox_cli.py [hunk 14]
```diff
@@ -141,12 +141,7 @@ def main():
     subcmd = commands[args.subcommand]()
     subcmd.register_arguments(subcmd_parser)
     sub_args = subcmd_parser.parse_args(sys.argv[subcmd_index + 1 :])
-    sa = SessionAssistant(
-        "com.canonical:checkbox-cli",
-        "0.99",
-        "0.99",
-        ["restartable"],
-    )
+    sa = SessionAssistant()
```
Comment (pieqq) [inline checkbox-ng/checkbox_ng/launcher/checkbox_cli.py:144]: I had been copy-pasting the previous blob of text every time I wanted to test something without realizing the parameters were not useful in most of the cases :S

**Review by pieqq — CHANGES_REQUESTED** [review]

File: checkbox-ng/checkbox_ng/launcher/agent.py (hunk 13 above)
Comment (kissiel) [reply checkbox-ng/checkbox_ng/launcher/agent.py]: Fixed and added tests for the case in 12583f244

File: checkbox-ng/checkbox_ng/launcher/test_agent.py (hunk 9 above)
Comment (kissiel) [reply checkbox-ng/checkbox_ng/launcher/test_agent.py]: Very valid point. Fixed.

File: checkbox-ng/plainbox/impl/session/remote_assistant.py (hunk 10 above)
Comment (kissiel) [reply checkbox-ng/plainbox/impl/session/remote_assistant.py]: Resurrected the call and mocked it in the UTs.

File: docs/explanation/remote.rst (hunk 12 above)
Comment (kissiel) [reply docs/explanation/remote.rst]: Understood. Trimmed the text to what you proposed.

File: checkbox-ng/plainbox/impl/session/test_remote_assistant.py (hunk 11 above)
Comment (kissiel) [reply checkbox-ng/plainbox/impl/session/test_remote_assistant.py]: We can. Changed for consistency. Thanks.

**Review by Hook25 — APPROVED** [review]
+1, well done!

File: checkbox-ng/checkbox_ng/launcher/checkbox_cli.py (hunk 14 above)
Comment (kissiel) [reply checkbox-ng/checkbox_ng/launcher/checkbox_cli.py:144]: Pierre told me on MM:

>oh sorry, it was just a comment for myself, mostly
>nothing to change

**Review by pieqq — APPROVED** [review]
Nice! Thanks!

**Review by Hook25 — APPROVED** [review]
+1

**Review by Hook25 — APPROVED** [review]
+1
